accessible/toolboxview.component.js

```diff
@@ -0,0 +1,139 @@
+            [ngClass]="{blocklyHasChildren: true,
+                blocklyActiveDescendant: tree.getAttribute('aria-activedescendant') == idMap['Parent' + i]}" 
+            *ngIf="toolboxHasCategories" *ngFor="#category of makeArray(sightedToolbox); #i=index" 
+            aria-level="1" aria-selected=false>
+          <!-- TODO(sll@): This div is required. There seems to be some bug in Angular that makes it 
+          access index=undefined. Talk to fraser@.-->
+          <div *ngIf="category && category.attributes">
+            <label [id]="idMap['Label' + i]" #name>{{category.attributes.name.value}}</label>
+            {{labelCategory(name, i, tree)}}
+            <ol role="group" *ngIf="getToolboxWorkspace(category).topBlocks_.length > 0">
+              <toolbox-tree-view *ngFor="#block of getToolboxWorkspace(category).topBlocks_" [level]=2 [block]="block" [displayBlockMenu]="true" [clipboardService]="clipboardService"></toolbox-tree-view>
+            </ol>
+          </div>
+        </li>
+        <div *ngIf="!toolboxHasCategories">
+          <toolbox-tree-view *ngFor="#block of getToolboxWorkspace(toolboxCategories[0]).topBlocks_; #i=index" [level]=1 [block]="block" [displayBlockMenu]="true" [clipboardService]="clipboardService" [index]="i" [tree]="tree" [noCategories]="true"></toolbox-tree-view>
+        </div>
+      </ol>
+    `,
+    directives: [blocklyApp.ToolboxTreeView],
+    providers: [blocklyApp.TreeService, blocklyApp.UtilsService],
+  })
+  .Class({
+    constructor: [blocklyApp.TreeService, blocklyApp.UtilsService, function(_treeService, _utilsService) {
+      this.sightedToolbox = document.getElementById('blockly-toolbox-xml');
+
+      this.toolboxCategories = [];
+      this.toolboxWorkspaces = Object.create(null);
+      this.treeService = _treeService;
+      this.utilsService = _utilsService;
+
+      this.toolboxHasCategories = false;
+    }],
+    ngOnInit: function() {
+      var elementsNeedingIds = [];
+      var categories = this.makeArray(this.sightedToolbox);
+      if (this.toolboxHasCategories) {
+        for (var i = 0; i < categories.length; i++){
+          elementsNeedingIds.push('Parent' + i);
+          elementsNeedingIds.push('Label' + i);
+        }
+        this.idMap = this.utilsService.generateIds(elementsNeedingIds);
+        this.idMap['Parent0'] = 'blockly-toolbox-tree-node0';
+      }
+    },
+    labelCategory: function(label, i, tree) {
+      var parent = label.parentNode;
+      while (parent && parent.tagName != 'LI') {
+        parent = parent.parentNode;
+      }
+      parent.setAttribute('aria-label', label.innerText);
+      parent.id = 'blockly-toolbox-tree-node' + i;
+      if (i == 0 && tree.getAttribute('aria-activedescendant') ==
+        'blockly-toolbox-tree-node0') {
+        this.treeService.setActiveDesc(parent, tree.id);
+        parent.setAttribute('aria-selected', 'true');
+      }
+    },
+    makeArray: function(val) {
```

Comment:
This has weird side effects. I don't expect calling makeArray() to set this.toolboxHasCategories.

Looking at how you use it, it seems you always call this with the same argument. Why not remove this method, transfer its contents to ngOnInit() (when you compute categories), set this.toolboxCategories to the result, and in the template refer to toolboxCategories instead? Perhaps I'm missing something, but I don't really understand why the categories need to be computed more than once.